services/LinkLoss/source/LinkLossService.cpp

```
@@ -0,0 +1,110 @@
+
+    _alert_level_char.setWriteAuthorizationCallback(this, &LinkLossService::onDataWritten);
+
+    ble_error_t error = _ble.gattServer().addService(linkLossService);
+
+    if (error == BLE_ERROR_NONE) {
+        _chainable_gap_event_handler.addEventHandler(this);
+    }
+}
+
+void LinkLossService::set_event_handler(EventHandler* handler)
+{
+    _alert_handler = handler;
+}
+
+void LinkLossService::set_alert_level(AlertLevel level)
+{
+    _alert_level = level;
+}
+
+void LinkLossService::set_alert_timeout(std::chrono::milliseconds timeout)
+{
+    _alert_timeout = timeout;
+}
+
+LinkLossService::AlertLevel LinkLossService::get_alert_level()
+{
+    return _alert_level;
+}
+
+void LinkLossService::stop_alert()
+{
+    _event_queue.cancel(_event_queue_handle);
+    _event_queue_handle = 0;
+    if (_in_alert) {
+        _in_alert = false;
+        if (_alert_handler) {
+            _alert_handler->on_alert_end();
+        }
+    }
+}
+
+void LinkLossService::onConnectionComplete(const ble::ConnectionCompleteEvent &event)
+{
+    if (event.getStatus() == BLE_ERROR_NONE) {
+        stop_alert();
+    }
+}
+
+void LinkLossService::onDisconnectionComplete(const ble::DisconnectionCompleteEvent &event)
+{
+    AlertLevel level = get_alert_level();
+    if (_alert_handler != nullptr && event.getReason() == ble::disconnection_reason_t::CONNECTION_TIMEOUT &&
+        level != AlertLevel::NO_ALERT && !_in_alert) {
+        _in_alert = true;
+        _alert_handler->on_alert_requested(level);
+        if (_alert_timeout > std::chrono::milliseconds(0)) {
+            _event_queue_handle = _event_queue.call_in(_alert_timeout, [this] { stop_alert(); });
+        }
+    }
+}
+
+void LinkLossService::onDataWritten(GattWriteAuthCallbackParams *write_request)
+{
+    const uint8_t level = *write_request->data;
+
+    if (level <= (uint8_t)(AlertLevel::HIGH_ALERT)) {
+        set_alert_level((AlertLevel) level);
+    }
```

Could you return an error if the alert is out of range ?